frontend/lib/src/components/shared/StreamlitMarkdown/StreamlitMarkdown.tsx

```diff
@@ -416,7 +416,36 @@ export function RenderedMarkdown({
           return
         }
 
-        // Handle color directives
+        // Handle badge directives (:color-badge[])
+        const badgeMatch = nodeName.match(/^(.+)-badge$/)
+        if (badgeMatch && colorMapping.has(badgeMatch[1])) {
+          const color = badgeMatch[1]
+
+          // rainbow-badge is not supported because the rainbow text effect uses
+          // background-clip: text with a transparent color, which conflicts with
+          // having a background color for the badge.
+          // We *could* support it by using a nested span structure, but that breaks
+          // the material icon handling below.
+          // We can support that in the future if we want to, but I think a
+          // rainbow-colored badge shouldn't be a common use case anyway.
+          if (color === "rainbow") {
+            return
+          }
+
+          const textColor = colorMapping.get(color)
+          const bgColor = colorMapping.get(`${color}-background`)
+
+          if (textColor && bgColor) {
+            const data = node.data || (node.data = {})
+            data.hName = "span"
+            data.hProperties = data.hProperties || {}
+            data.hProperties.className = "has-background-color"
+            data.hProperties.style = `${bgColor}; ${textColor}; font-size: ${theme.fontSizes.sm}; white-space: nowrap;`
```

Comment:
What happens if the badge is longer than the container width?

Reply:
Hm it overflows currently. I tried making it in a way where it doesn't wrap by default but does wrap if it's longer than one line, but can't really get it to work (only with using `display: inline-block` but that seems to add a tiny bit of extra padding in the badge which looks awkward). Do you have any idea how to get this to work? 

Otherwise, we need to decide if we either want to avoid very long badges overflowing, or avoid small badges wrapping. I'm slightly leaning towards the latter since very long badges don't seem like a very common use case.

Reply:
> we need to decide if we either want to avoid very long badges overflowing, or avoid small badges wrapping. I'm slightly leaning towards the latter since very long badges don't seem like a common use case to me.

Hmm, yeah I would probably also prefer the second case but I think it could create quite a bit of bad-looking UI in situations where the badge is used within containers with a smaller width (e.g. columns). Maybe we can set a reasonable max width and ellipsis or scrolling on overflow?

Reply:
OK I made it work with inline-block by changing the padding a bit. And also made it ellipsize.

Reply:
Awesome, maybe add one badge to the e2e example with a long label to trigger the behavior?